### PR TITLE
Macros3

### PR DESCRIPTION
**Pull request by fponticelli**

### CODE REVIEW

File: src/thx/schema/Core.hx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      thx.Url.fromString,
+      (v: thx.Url) -> v.toString()
+    );
+
+  public static function intMap<T>(elementSchema: Schema<String, T>): Schema<String, Map<Int, T>>
+    return unsafeParseSchema(
+      thx.schema.SimpleSchema.dict(elementSchema),
+      function(m: Map<String, T>): Map<Int, T> {
+        var imap = new Map();
+        for(k in m.keys()) {
```
Comment (nuttycom) [inline src/thx/schema/Core.hx]: Do you maybe want to use `.traverseValidation` here so that you can collect all the keys that do not parse correctly, rather than just the first, and use a normal parse schema?

File: src/thx/schema/SimpleSchema.hx [hunk 2]
```diff
@@ -66,6 +64,12 @@ class SimpleSchema {
       constAlt("none", None, function(a: Option<A>, b: Option<A>) return a == b)
     ]);
 
+  public static function makeNullable<E, A>(s: Schema<E, A>): Schema<E, Null<A>>
+    return oneOf([
+      alt("value", s, function(a: A) return a, function(a: Null<A>) return Options.ofValue(a)),
+      constAlt("novalue", null, function(a: A, b: A) return a == b)
```
Comment (nuttycom) [inline src/thx/schema/SimpleSchema.hx:70]: Is `==` the correct thing to use here? Universal equality is a lie; is this just a reference equality check or something?

File: src/thx/schema/macro/AnonField.hx [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package thx.schema.macro;
+import haxe.macro.Context;
+import haxe.macro.Type;
+
+class AnonField {
+  public static function fromEnumArg(arg: { t: Type, opt: Bool, name: String })
+    return new AnonField(arg.name, BoundSchemaType.fromType(arg.t));
+
+  public var name: String;
```
Comment (nuttycom) [inline src/thx/schema/macro/AnonField.hx]: `(default, null)`

File: src/thx/schema/macro/Arguments.hx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import haxe.macro.Context;
+import haxe.macro.Expr;
+import thx.schema.macro.Error.*;
+import haxe.ds.Option;
+
+class Arguments {
+  /**
+   *  Generates a schema from the type passed as the first argument.
+   *  @param typeRef
```
Comment (nuttycom) [inline src/thx/schema/macro/Arguments.hx]: Is this comment wrong? The array of exprs is checked to have length no greater than 2.

File: src/thx/schema/macro/Arguments.hx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+   *  @param ?typeSchemas
+   *  @param ?identifier
+   */
+  public static function parseArguments(exprs: Array<Expr>) {
```
Comment (nuttycom) [inline src/thx/schema/macro/Arguments.hx]: Please explicitly ascribe return types.

File: src/thx/schema/macro/Arguments.hx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    function normalizeExpr(e) {
+      return Context.getTypedExpr(Context.typeExpr(e));
+    }
+    switch typeSchemas.expr {
+      case EConst(CIdent("null")):
+      case EArrayDecl(arr):
+        for(item in arr) {
+          switch Context.typeof(item) {
+            case TType(_.toString() => stype, [_, t]) if(stype == "thx.schema.Schema"):
```
Comment (nuttycom) [inline src/thx/schema/macro/Arguments.hx]: `thx.schema.Schema` isn't an actual type. Is this supposed to be thx.schema.SchemaF?

File: src/thx/schema/macro/BoundSchemaType.hx [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    } catch(e: Dynamic) {
+      None;
+    }).map(fromType);
+  }
+
+  public static function createLocalParam(param: String)
+    return new BoundSchemaType(LocalParam(param));
+
+  public static function fromTypeOrTypeParam(type: Type, params: Array<BoundSchemaType>) {
+    var names = params.map(p -> p.toString());
```
Comment (nuttycom) [inline src/thx/schema/macro/BoundSchemaType.hx]: `names` is not used in this function

File: src/thx/schema/macro/BoundSchemaType.hx [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    function resolve(t: Type) return switch t {
+      case TLazy(f): f();
+      case other: other;
+    }
+    type = resolve(type);
```
Comment (nuttycom) [inline src/thx/schema/macro/BoundSchemaType.hx]: `return fromType(resolve(type))` is surely a nicer way to write this. Also ascribing the return type of `resolve` would be helpful for future people trying to understand this.

File: src/thx/schema/macro/BoundSchemaType.hx [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+      case TType(_.get() => t, p):
+        fromDefType(t, p);
+      case TMono(_.get() => t):
+        fromType(t);
+      case TFun(_):
+        fatal('A schema for the function `${TypeTools.toString(type)}` cannot be generated. Are you trying to generate a schema for an anonymous object that contains a method?');
+      case TLazy(f):
+        fromType(f());
+      case _:
+        throw 'Unable to convert type to BoundSchemaType: $type';
```
Comment (nuttycom) [inline src/thx/schema/macro/BoundSchemaType.hx]: What is the difference between `throw` and `fatal` in this function?

File: src/thx/schema/macro/BoundSchemaType.hx [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+import haxe.macro.Type;
+import haxe.macro.TypeTools;
+import thx.schema.macro.Error.*;
+import thx.schema.macro.Utils.*;
+import haxe.ds.Option;
+using thx.Options;
+
+// this is the type for things inside object literals used like this: Generic.schema({ name: Option<String> })
+// and also it is used when looking up schemas during generation
+class BoundSchemaType {
```
Comment (nuttycom) [inline src/thx/schema/macro/BoundSchemaType.hx:15]: It looks to me like a lot of this machinery is not directly related to the schema algebra. Is this stuff that could actually be useful as part of a more general haxe metaprogramming library?

File: src/thx/schema/macro/SchemaBuilder.hx [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        generateAbstractSchema(abs, qtype, typeSchemas);
+      case [TAnonymous(_.get() => anon),  BoundSchemaTypeImpl.AnonObject(obj)]:
+        generateAnonSchema(anon,    obj,   typeSchemas);
+      case [TType(_.get() => def, _),  BoundSchemaTypeImpl.QualifiedType(qtype)]:
+        generateDefTypeSchema(def, qtype,  typeSchemas);
+      case _: fatal('Cannot generate schema for unsupported type ${schemaType.toString()}');
+    }
+  }
+
+  static function generateClassSchema(cls: ClassType, qtype: QualifiedType<BoundSchemaType>, typeSchemas: Map<String, Expr>) {
```
Comment (nuttycom) [inline src/thx/schema/macro/SchemaBuilder.hx:69]: Something that I've found very useful in a number of places is, instead of creating an `object(...)` schema, just creating the `ObjectBuilder` value that the `ObjectSchema` wraps. The advantage is that `ObjectBuilder` values compose. Is that something that this macro library could support as well?

File: src/thx/schema/macro/SchemaBuilder.hx [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+    return if(n == 0) {
+      var path = qtype.parts();
+      macro thx.schema.SimpleSchema.object(PropsBuilder.Pure(Type.createEmptyInstance($p{path})));
+    } else {
+      // generate constructor function
+      var constructor = generateClassConstructorF(fields.map(classFieldToFunctionArgument), qtype);
+      // generate fields
+      var properties = fields.map(createPropertyFromClassField.bind(BoundSchemaType.createQualified(qtype), typeSchemas, _));
+      // capture apN and ap arguments
+      var apN = 'ap$n';
```
Comment (nuttycom) [inline src/thx/schema/macro/SchemaBuilder.hx:81]: You should probably fail before here if there's no implementation of `ap$n` available for that `$n`, unless you want to also generate the required `ap` function (which should be straightforward).

File: src/thx/schema/macro/SchemaBuilder.hx [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+
+  static function generateClassSchema(cls: ClassType, qtype: QualifiedType<BoundSchemaType>, typeSchemas: Map<String, Expr>) {
+    var fields = cls.fields.get().filter(keepVariables);
+    var n = fields.length;
+    return if(n == 0) {
+      var path = qtype.parts();
+      macro thx.schema.SimpleSchema.object(PropsBuilder.Pure(Type.createEmptyInstance($p{path})));
+    } else {
+      // generate constructor function
+      var constructor = generateClassConstructorF(fields.map(classFieldToFunctionArgument), qtype);
```
Comment (nuttycom) [inline src/thx/schema/macro/SchemaBuilder.hx:77]: Since in Haxe constructors are used to enforce object invariants, this seems like it might create a bit of a hole in applications where deserialization could construct "illegal" objects leading to undefined behavior. I know it's more work, but what if for classes and abstracts one were to type-infer which constructor to use based upon the types of the schema arguments passed, and then always route through those constructors?

File: src/thx/schema/macro/SchemaBuilder.hx [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+      case [TAbstract(_.get() => t, p), BoundSchemaTypeImpl.QualifiedType(qtype)]:
+        generateAbstractSchema(t, qtype, typeSchemas);
+      case [_, s]:
+        fatal('Unsupported Type Definition for type: ${def.type} and $s');
+    }
+  }
+
+  static function generateAnonSchema(anon: AnonType, anonObject: AnonObject<BoundSchemaType>, typeSchemas: Map<String, Expr>) {
+    var n = anonObject.fields.length;
+    var apN = 'ap$n';
```
Comment (nuttycom) [inline src/thx/schema/macro/SchemaBuilder.hx:172]: Same concern about `n` here as above.

File: src/thx/schema/macro/SchemaBuilder.hx (hunk 13 above)
Comment (nuttycom) [reply src/thx/schema/macro/SchemaBuilder.hx:77]: I guess maybe this doesn't really work, as there's no way to invert those constructor functions to discover how the fields on the object map back to the constructor arguments for serialization. Bummer.

File: test/thx/schema/TestGeneric.hx [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+  // O<T3>(t3: Option<Array<T3>>);
+}
+
+enum Case2 {
+  A;
+  B(di: MyInt, z: haxe.ds.Option<String>);
+}
+
+enum GADT<T> {
+  A(a: T): GADT<Int>;
```
Comment (nuttycom) [inline test/thx/schema/TestGeneric.hx:382]: Just a general question here - does `T` unify with `Int` in this case (and `String`) below? I haven't seen a GADT type variable in this position before.

**Review by nuttycom — COMMENTED** [review]
Overall this looks pretty amazing. It would benefit from a few more comments, and I wonder how much of it is specific to schema vs. how much could be factored out to a more general-purpose library. Unfortunately I'm not really qualified to assess most of the macro code, but the tests look good so 👍

File: src/thx/schema/SimpleSchema.hx (hunk 2 above)
Comment (fponticelli) [reply src/thx/schema/SimpleSchema.hx:70]: True, structural equality can be achieved with `thx.Dynamcs.equals`. Will do that. It is also true that we are comparing to a `null` value so strict equality should always work?

File: test/thx/schema/TestGeneric.hx (hunk 15 above)
Comment (fponticelli) [reply test/thx/schema/TestGeneric.hx:382]: It definitely should but this is not implemented yet. In fact the test case is commented out :)

File: src/thx/schema/macro/SchemaBuilder.hx (hunk 12 above)
Comment (fponticelli) [reply src/thx/schema/macro/SchemaBuilder.hx:81]: `ap` is not generated (it would be nice). In case of missing `ap` the compiler should say exactly that. The error might non make much sense to the unaware user. If we put a check for `n` limit it means that when we add more `apX` we have 2 places to check. Ideally the generation of `ap` would solve that.

File: src/thx/schema/macro/SchemaBuilder.hx (hunk 13 above)
Comment (fponticelli) [reply src/thx/schema/macro/SchemaBuilder.hx:77]: Yeah, this is a tough one. We might annotate the type with a function that does that but then we move the responsibility of consistency to the user. The reflection way is dirty but should work in most cases ... it captures all fields, private and public.

File: src/thx/schema/macro/SchemaBuilder.hx (hunk 11 above)
Comment (fponticelli) [reply src/thx/schema/macro/SchemaBuilder.hx:69]: We could factor that out, at this time the only exposed function is `thx.schema.Generic.schema`.

File: src/thx/schema/macro/BoundSchemaType.hx (hunk 9 above)
Comment (fponticelli) [reply src/thx/schema/macro/BoundSchemaType.hx]: I should use fatal :)

File: src/thx/schema/macro/BoundSchemaType.hx (hunk 10 above)
Comment (fponticelli) [reply src/thx/schema/macro/BoundSchemaType.hx:15]: Yeah potentially. The `UnboundSchemaType` is probably not even necessary and I have to try to remove it.

**Comment by fponticelli** [comment]
Ok, I will come back with fixes to the notes above.

File: src/thx/schema/SimpleSchema.hx (hunk 2 above)
Comment (fponticelli) [reply src/thx/schema/SimpleSchema.hx:70]: My understanding here is that this equality function is always going to compare with the `const` value (`null`) ... so strict equality is fine? You do the same just above in `makeOptional`.

File: src/thx/schema/macro/Arguments.hx (hunk 4 above)
Comment (fponticelli) [reply src/thx/schema/macro/Arguments.hx]: No it is actually correct. Optional arguments are not easy in macro land and it is easier to check the arity yourself by using the magic `Array<Expr>`

File: src/thx/schema/macro/Arguments.hx (hunk 6 above)
Comment (fponticelli) [reply src/thx/schema/macro/Arguments.hx]: That's actually how the compiler sees it. It does exist as a type alias: `thx.schema.SimpleSchema.Schema`.

File: src/thx/schema/SimpleSchema.hx (hunk 2 above)
Comment (nuttycom) [reply src/thx/schema/SimpleSchema.hx:70]: Okay, sounds good.

**Comment by fponticelli** [comment]
Ok, @nuttycom I applied the changes that made sense
